src/utils/jsonlogic.ts

```diff
@@ -0,0 +1,169 @@
+  formDefinition: string | null;
+  name: string;
+  key: string;
+  source:
+    | '' // static
+    | 'component'
+    | 'user_defined';
+  serviceFetchConfiguration: ServiceFetchConfiguration | null;
+  prefillPlugin: string;
+  prefillAttribute: string;
+  prefillIdentifierRole: 'authorised_person' | 'main';
+  dataType: DataType; // (incomplete) type annotation
+  dataFormat: string;
+  isSensitiveData: boolean;
+  initialValue: JSONValue;
+}
+
+interface FormContext {
+  formVariables?: VariableDefinition[];
+  staticVariables?: VariableDefinition[];
+  components?: AnyComponentSchema[];
+}
+
+/**
+ * Return a function that can type check whether
+ */
+export const createTypeCheck = ({
+  formVariables = [],
+  staticVariables = [],
+  components = [],
+}: FormContext = {}): JsonLogicTypeChecker => {
+  const variableContext = staticVariables.concat(formVariables);
+  let data: JSONObject = Object.fromEntries(
+    variableContext.map(variable => [variable.key, dataTypeForVariableDefinition(variable)])
+  );
+  // formVariables contain user defined and component variables
+  // but dataType `array` and `object` are incomplete
+  // overwrite the component variables
+  Object.entries(components).map(
+    ([key, component]) => (data[key] = dataTypeForComponent(component))
+  );
+
+  return (logic, expected = undefined) => {
+    // We don't evaluate logic, we just look at types.
+    // "===" forces the logic expression align with the expectancy if defined
+    const result = infer(expected !== undefined ? {'===': [logic, expected]} : logic, data);
+    return result.startsWith('result type:') ? '' : result;
+  };
+};
+
+// TODO use infer if serviceFetchConfiguration.data_mapping_type == 'JsonLogic'
+const dataTypeForVariableDefinition = ({initialValue, dataType}: VariableDefinition): JSONValue =>
+  initialValue !== null
+    ? initialValue // may be something more complete than [] or {} ... TODO prefillAttribute too?
+    : {
+        string: '',
+        int: 1,
+        float: 1.1,
+        array: [],
+        object: {},
+        date: '',
+        datetime: '',
+        time: '',
+        boolean: true,
+      }[dataType];
+
+const dataTypeForComponent = (
+  component:
+    | AnyComponentSchema
+    | {
+        type: 'map' | 'editgrid' | 'password' | 'signature'; // Not yet implemented in this builder
+        multiple?: boolean;
+        defaultValue: JSONValue;
+      }
+): JSONValue => {
+  // For now return example values as accepted by InferNoLogic
+  // But example values cannot distinguish arrays from tuples!
+  const value = {
+    address: {
+      postcode: '',
+      houseNumber: '',
+      houseLetter: '',
+      houseNumberAddition: '',
+    },
+    currency: 1,
+    number: 1,
+    checkbox: true,
+    //@ts-ignore selectboxes always have component.defaultValue (this does work when rewritten as a lengthy switch/case)
+    selectboxes: component.defaultValue,
+    npFamilyMembers: {}, // TODO record type
```

Comment:
this ultimately becomes the same as selectboxes BUT it can also become a `type: 'content'` component if there's an error :grimacing: